### ISSUE TITLE
'Mountable' widgets

### ISSUE THREAD

**Issue by maartenbreddels** [jupyter-widgets/ipywidgets#2410]
# Concept
Using voila, or embedding DOMWidgets into a html page I may want to specify a particular id for a widget, and on the page, have a way of specifying where that widget should go. To make it more concrete:
```python
button = Button(description='Logout', mount_id='logout')
```
And in the placing html:
```html
<....>
<div id='logout'>
 ... some possible placeholder until the widgets gets loaded...
</div>
```

A custom widget manager may then inspect all DOMWidget, and for those that have a mount_id, create a view and place them at a html element with that id. 

Note that this is a specific example, but adding a `mount_id = Unicode()` trait to the DOMWidget is a requirement for this.
React and Vue based WidgetManagers may have their own opinionated ways of 'mounting' widgets to specific components/location, but the use of a string is quite general.

# Proposal
Add a `mount_id = Unicode()` trait to the DOMWidget for custom placement of widgets onto webpages in voila or other web context.

**Comment by jasongrout** [comment]
This breaks the idea that widgets can have multiple views, though.

The widget id is already unique, and can be specified (IIRC). If you want the above, with the constraints of only one widget on a page, could you just give a list of widget ids to display to the widget manager, and then it goes and displays them?

**Comment by maartenbreddels** [comment]
I agree, with the example I give it seems to only allow 1 view. However, this is simply a usage example, in React, you can imagine a `<WidgetMountPoint mount_id='foobar'/>`, which could occur multiple times.
What is new to me is that the widget id can be set. Do you suggest using this, in order to give widgets a human readable id? 

I think it can be really powerful with a voila template that does not show cell output, but instead 'mounts' widgets into pre-named positions in the template.

**Comment by SylvainCorlay** [comment]
The `AppLayout` widget is a good example of a container widget with named areas.

Maybe a solution would be to allow users to make a new container with named areas in pure HTML?

**Comment by SylvainCorlay** [comment]
That HTML template could be both used as a widget container and a portion of a voila template.

**Comment by maartenbreddels** [comment]
You can look at it as a container with named areas, but that's already an implementation specific idea.

 I think it could be very useful to have an implementation agnostic name, that helps frontends position specific widgets in 1 (or more) places. Even the mount_id does not have to be unique (maybe id is then not the right word, mount_point/name/placement/..), another implementation might collect all widgets with mount_id='toolbar' in a toolbar that has no specific id or className?

Some other ideas, how this can be useful:
 * Jupyter lab plugin that places all widgets with a specific mount_id in the toolbar/menu/wherever.
 * A React voila template that has a navigation drawer, that places all widgets with 'navigation_drawer' in it (attached to the React vdom).
 * A jupyter server extension, that list on a page, for each running kernel/session the widgets that have a mount_id/mount_point/name and allows you to open it in a new tab (voila based). A concrete example would be an ipyvolume widget, that you created in your notebook, and want to have (live, connected) opened on your mobile phone, so you can do some VR exploration with a connected kernel.

Let me refine my proposal a bit:

Add a Unicode trait (mount_id/name/widget_name/placement/...) that allows users to give human readable (non-unique) names/ids to widgets. Implementations can use this for placements, for listing, or other not-thought of ways.

**Comment by jasongrout** [comment]
Why stop at a single name? Would it be useful to have multiple tags for a widget? Like `['toolbar', 'left', 'main', 'display-on-hover']`?

**Comment by maartenbreddels** [comment]
I like it, also the name tags would be more neutral.

(from mobile phone)

**Comment by maartenbreddels** [comment]
Refined proposal (#3):
add a `tags = List(String)` trait to the core widget for user-defined purposes.

**Comment by jasongrout** [comment]
One possibly confusing collision: traits metadata is added with the `.tag()` method (like `.tag(sync=True)`).

**Comment by maartenbreddels** [comment]
True, I also thought of `metadata`, but that implies any kind of data, and actually has the same issue!. On the other hand, `.tag()` is only used by library developers, users would only be exposed to the `tags` trait.

**Comment by vidartf** [comment]
Related PR? https://github.com/jupyter-widgets/ipywidgets/pull/2431

**Comment by martinRenou** [comment]
I think we (@maartenbreddels and I) came to a good solution on Gitter about this issue. Instead of adding metadata for making the widget mountable on a specific place on the page, we could use the `display_id` from the IPython `display` function.

So:
```python
button = Button(description='Logout', mount_id='logout')
```
would become:
```python
button = Button(description='Logout')
display(button, display_id='logout')
```

The only issue we saw with this approach is that the `display_id` is considered a "transient" information (see https://jupyter-client.readthedocs.io/en/stable/messaging.html#display-data). So it does not persist in the Notebook.

**Comment by maartenbreddels** [comment]
After that, we found the metadata argument of display, which allows anything:
https://ipython.org/ipython-doc/dev/api/generated/IPython.display.html#IPython.display.display

I do prefer the solution we use in ipyvuetify using a trait because we don't always want to display widgets, some are only created for the voila dashboard.

**Comment by martinRenou** [comment]
> After that, we found the metadata argument of display, which allows anything

Right, I forgot about this. Indeed it would be a good solution too.

> I do prefer the solution we use in ipyvuetify using a trait because we don't always want to display widgets, some are only created for the voila dashboard.

Good point.
But I have to say I prefer the `IPython.display` solution, because sometimes you don't want to display a widget but a PyViz plot for example. It is not really an Interactive Widgets issue but more of a cell output issue.

**Comment by maartenbreddels** [comment]
I agree, both solutions are needed to support non-widgets, and we plan to support both in voila-vuetify.

**Comment by martinRenou** [comment]
So do you think it is fine to close this issue or do you still think this should require a change in ipywidgets?

**Comment by maartenbreddels** [comment]
I'm not sure we should close it since I don't think we've found the right solution yet. People don't look at closed issues that much.
